update.go

```diff
@@ -254,6 +260,17 @@ other options are ignored.
 		config.Cgroups.Resources.MemorySwap = *r.Memory.Swap
 		config.Cgroups.Resources.PidsLimit = r.Pids.Limit
 
+		if val := context.String("l3-cache-schema"); val != "" {
+			// Avoid silent failure in container.Set()
+			if !intelrdt.IsEnabled() || config.IntelRdt == nil {
```

Comment:
Why  would `config.IntelRdt == nil` be a case for an error?

Reply:
@crosbymichael 
>Why would config.IntelRdt == nil be a case for an error?

Sorry, this is not straightforward.

1. If linux.intelRdt is not specified in config.json, always config.IntelRdt == nil (https://github.com/opencontainers/runc/blob/master/libcontainer/specconv/spec_linux.go#L253-L258)
2. In update.go, config := container.Config() (https://github.com/opencontainers/runc/blob/master/update.go#L150), we get `config.IntelRdt == nil`, which implies linux.intelRdt is not specified in config (even though intelrdt.IsEnabled() == true).
3. In my understanding, this is an error case in updateCommand because intelRdt is not managed in container processes management life cycle - no Apply(), Destroy() and etc., only invoke container.Set() --> IntelRdtManager.Set() in updateCommand.

Reply:
That sounds like a bug to me. If someone wants to add an intelrdt limit after the fact (where they didn't specify any options in the original configuration) that should be possible -- is this a kernel bug or just a limitation of `runc` at the moment? We allow you to do that with all of the other cgroups from memory.

Reply:
@cyphar 

> That sounds like a bug to me. If someone wants to add an intelrdt limit after the fact (where they didn't specify any options in the original configuration) that should be possible -- is this a kernel bug or just a limitation of `runc` at the moment? We allow you to do that with all of the other cgroups from memory.

This is what I intended to do for Intel RDT/CAT support in `runc` as a **tradeoff**. In my opinion, this is neither a kernel bug nor a runc limitation. Let me explain here:

Unlike cgroups' hierarchy, Intel RDT resource control filesystem in kernel supports only single level filesystem layout. (see details in section "Intel RDT "resource control" filesystem hierarchy" in #433 ). For Cache Allocation Technology (CAT) feature, it only supports limited number of groups. It is hardware limitation by nature. The limitation is indicated in /sys/fs/resctrl/info/L3/num_closids. 

In `runc` case, we could only create limited number <container_id> directories under /sys/fs/resctrl via `Apply()`. So we need a **tradeoff**: we `Apply()` intelrdt constraint on a container **only** if hardware and kernel support Intel RDT feature and intelRdt is specified in original configuration. Otherwise, if hardware and kernel support the feature, we will run out of the limitation quickly when started a lot of containers instances, even though the user doesn't care about Intel RDT/CAT feature or doesn't want to add intelrdt constraint for some container instances.

Reply:
@xiaochenshen Alright, that makes sense (as in why it's not that way at the moment) -- but that just sounds like we need code to create a new `rdtgroup` if we're doing an update? I think just using `Apply` might be enough, but I would have to check the code again.

Reply:
@cyphar 
> but that just sounds like we need code to create a new rdtgroup if we're doing an update? I think just using Apply might be enough, but I would have to check the code again.

I will have a look how to do it in update command. Thank you for the suggestion.